Sources/Aardvark/BugReporting/FileSystemAttachmentGenerator.swift

```diff
@@ -0,0 +1,219 @@
+//  Copyright 2021 Square, Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+@objc(ARKFileSystemAttachmentGenerator)
+public final class FileSystemAttachmentGenerator: NSObject {
+
+    // MARK: - Public Static Methods
+
+    /// Generates bug report attachments showing a breakdown of the file system.
+    ///
+    /// The attachment consists of:
+    /// * A list of files in each of the specified `searchPathDirectories` with their file size and content modification
+    ///   date.
+    /// * Information about the total and available capacity on the main device volume.
+    ///
+    /// - parameter searchPathDirectories: The search paths within the app's container that should be scanned for files.
+    public static func attachment(
+        searchPathDirectories: [FileManager.SearchPathDirectory] = [
+            .documentDirectory,
+            .applicationSupportDirectory,
+            .cachesDirectory,
+        ]
+    ) throws -> ARKBugReportAttachment {
+        enum Error: Swift.Error {
+            case missingDocumentDirectory
+            case missingVolumeResourceValues
+        }
+
+        let fileManager = FileManager.default
+
+        let urls = searchPathDirectories.reduce([]) {
+            $0 + fileManager.urls(for: $1, in: .userDomainMask)
+        }
+
+        var description = ""
+
+        // We need an arbitrary URL on the main device volume in order to check the capacity and determine the URL for
+        // the container. We'll use the document directory since it should always exist.
+        guard let volumeURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
+            throw Error.missingDocumentDirectory
+        }
+
+        let appContainerPrefixes = try self.appContainerPrefixes(for: urls.first!)
+
+        if let primaryContainerPath = appContainerPrefixes.first {
+            description += "Showing files from search paths in app container at \(primaryContainerPath)\n\n"
+        }
+
+        let propertyKeys: [URLResourceKey] = [
+            .fileSizeKey,
+            .contentModificationDateKey,
+        ]
+
+        let fileSizeFormatter = ByteCountFormatter()
+        fileSizeFormatter.zeroPadsFractionDigits = true
+        fileSizeFormatter.allowsNonnumericFormatting = false
+        fileSizeFormatter.formattingContext = .listItem
+        fileSizeFormatter.isAdaptive = true
+
+        let maxPaddedFileSizeCharacterLength = 12
```

Comment:
Would be good to have a comment explaining the magic number

Reply:
Addressed in 6632635